[Lock] Added Symfony/Component/Lock/Store/MongoDbStore

kralos opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes (requires `ext-mongodb` and `mongodb/mongodb` to test)
| Fixed tickets | #27345 
| License       | MIT
| Doc PR        | symfony/symfony-docs#9807

**Testing caveat**  
In order to test this, the test environment needs `ext-mongodb` and `mongodb/mongodb`. Travis has `ext-mongodb` but Appveyor doesn't meaning I can't `require-dev` `mongodb/mongodb`. I had a look at `Symfony\Component\HttpFoundation\Session\Storage\Handler\MongoDbSessionHandler` since it poses the same issue and neither `symfony/symfony/src/Symfony/Component/HttpFoundation/composer.json` nor `symfony/symfony/composer.json` `require-dev` `mongodb/mongodb`. They both skip mongo based tests in the ci environments. I have therefore chosen to omit `require-dev` `mongodb/mongodb`.

I have both written the test and tested `Symfony\Component\Lock\Store\MongoDbStore` and it does pass in an environment with `ext-mongodb` and `mongodb/mongodb`.

**Description**  
We should support Semaphore Locks with a MongoDB back end to allow those that already use MongoDB as a distributed storage engine.

Symfony already partially supports MongoDB for session storage: `Symfony\Component\HttpFoundation\Session\Storage\Handler\MongoDbSessionHandler`

**Example**

```php
$client = new MongoDb\Client();

$store = new Symfony\Component\Lock\Store\MongoDbStore(
    $client
    array(
        'database' => 'my-app',
    )
);
$lockFactory = new Symfony\Component\Lock\Factory($store);
$lock = $lockFactory->createLock('my-resource');
```

Commit:
#27345 Added Symfony/Component/Lock/Store/MongoDbStore

Commit:
#27345 Fixed typo

Commit:
#27345 fixed coding standards

kralos commented:
@nicolas-grekas what's the deal with adding `ext-mongodb` to `symfony/symfony`'s appveyor build environment? Could you help me out?

nicolas-grekas commented:
No need for the ext on appveyor. Just skip tests there. An `@requires extension mongodb`annotation on the test class should do it.

Commit:
#27345 Removed dev requirement for mongodb/mongodb

travis env has ext-mongodb but appveyor doesn't

since MongoDbSessionHandler does the same i've stripped the dev requirement for mongodb/mongodb

Commit:
#27345 Removed unnessasary reduceLifetime

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+     * collection: The name of the collection [default: lock]
+     * resource_field: The field name for storing the lock id [default: _id] MUST be uniquely indexed if you chage it
+     * token_field: The field name for storing the lock token [default: token]
+     * acquired_field: The field name for storing the acquisition timestamp [default: acquired_at]
```
what is the purpose of this field?

kralos commented:
It's just to help show when the lock was first acquired, I can remove it if
you think we shouldn't be storing this? I was using it during development
to ensure locks were being cleaned up properly.

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+            $this->options['resource_field'] => (string) $key,
+            $this->options['token_field'] => $this->getToken($key),
+            $this->options['expiry_field'] => array(
+                '$gte' => $this->createDateTime(),
```
I'd rather define a `now` variable at the beggining of the method, then use this variable as a reference for every computes

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+        $key->reduceLifetime($ttl);
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
+        } catch (\MongoDB\Driver\Exception\BulkWriteException $e) {
```
same here

kralos replied:
improved exception handling as mentioned above

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+        try {
+            $result = $this->getCollection()->deleteOne($filter);
+        } catch (\MongoDB\Driver\Exception\BulkWriteException $e) {
+            throw new LockConflictedException('Failed to delete lock', 0, $e);
```
deletion does not throw ConflictException

kralos replied:
Fixed

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+     * garbage-collection. Alternatively it's possible to automatically expire
+     * the locks in the database as described below:
+     *
+     * A TTL collections can be used on MongoDB 2.2+ to cleanup expired locks
```
The current implementation does not purge expired locks. I'd rather use the verb MUST BE used,

kralos replied:
changed

jderusse reviewed (commented):
I'm not familiar with MongoDb, I'm OK with the current implementation (with few comments), 

BUT I've many concerns about the reliability of MongoDb. It's wellknown for not beeing ACID, whereas Atomicity is a requirements. What's about a MongoDb cluster configured with sharding and replication?

kralos commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+
+        $key->reduceLifetime($this->initialTtl);
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
```
MongoDB is atomic on a singular statement / document only. Hence the use of a update with filter (1 statement). MongoDB `^4.0` introduces multi statement transactions however this implementation does not rely on that. It should be atomic on all versions. Replication propagation can be controlled with a statement level options called a write concern. I'm adding support for that now. By default, mongo will read and write only to the `PRIMARY` node in a replica set (this is equivalent of reading and writing only to the master of a replicated RDBMS with replication slaves).

kralos replied:
With regards to `writeConcern`, `readConcern` and `readPreference` I've opted to leave this to the collection fallback placing the onus on the DBA to setup the collection with alternate settings if desired. The default when left unspecified is to read / write ONLY to the PRIMARY of a replica set. This is safe for the use of locks (it's like only reading / writing to a master in an RDBMS) however an alternate strategy could be left to the DBA depending on the number of nodes they have.

I have added some documentation regarding this decision to `symfony-docs`

kralos commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+     * collection: The name of the collection [default: lock]
+     * resource_field: The field name for storing the lock id [default: _id] MUST be uniquely indexed if you chage it
+     * token_field: The field name for storing the lock token [default: token]
+     * acquired_field: The field name for storing the acquisition timestamp [default: acquired_at]
```
I've removed it. I found it useful to debug / see when a lock was acquired. Probably not that useful in operation.

kralos commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+     * the locks in the database as described below:
+     *
+     * A TTL collections can be used on MongoDB 2.2+ to cleanup expired locks
+     * automatically. Such an index can for example look like this:
```
I've added a comment to the phpdoc and `symfony-docs`. I also recommended setting the `expireAfterSeconds` to a value greater than 0. The actual expiry is already handled by the application (not the DBMS's clocks). I have updated my example to 60 seconds.

jderusse replied:
IMHO this safety window should be a constructor parameter. Because the same issue exist on the `save` method when 2 clients are not  time synchronized

kralos replied:
Currently the recommendation is to manually create the TTL index. If I were to implement this as a constructor parameter in code we would be creating / updating the index on every execution. I'm not sure this is a good idea.

jderusse replied:
I would simply set the TTL to 0, and append this "clockDrift" parameter to the expiredAt (expiredAt = now + lockTTL + clockDrift). It would both solve the purge issue by Mongo and the race condition by 2 clients.

kralos replied:
I've moved clock discrepancy handling to a constructor option called `drift` instead of relying on a TTL index delay.

kralos commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+        $expiry = $this->createDateTime(microtime(true) + $this->initialTtl);
+
+        $filter = array(
+            $this->options['resource_field'] => (string) $key,
```
Yes, 1024 bytes including structure overhead (e.g. int / string) I've added this to `symfony-docs` under the Reliability section and to the phpdoc

kralos commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+        $update = array(
+            '$set' => array(
+                $this->options['resource_field'] => (string) $key,
+                $this->options['token_field'] => $this->getToken($key),
```
fixed

kralos commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,294 @@
+        $key->reduceLifetime($this->initialTtl);
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
+        } catch (\MongoDB\Driver\Exception\BulkWriteException $e) {
```
I've improved the exception handling, I also noticed `waitAndSave` was throwing an `InvalidArgumentException` instead of a `NotSupportedException`. The `RedisStore` also had this issue. I have fixed the `MondoDbStore` but we should also fix the `RedisStore`. I also adjusted the phpdoc in `StoreInterface` to include `LockExpiredException` in `save`, `waitAndSave` and `putOffExpiration` as per other store implementations.

Commit:
#27345 Improved documentation, exception handling, removed configurable fields

Commit:
#27345 fixed typo and @param indentation

Commit:
#27345 fixed @param indentation

Commit:
#27345 moved clock discrepancy handling to a constructor option 'drift'

Commit:
#27345 fixed phpdoc spacing

kralos commented:
@jderusse should be good for a second look. Thanks for you feedback so far, really helpful.

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,263 @@
+     * must not exceed 1024 bytes including structural overhead.
+     * @see https://docs.mongodb.com/manual/reference/limits/#Index-Key-Limit
+     *
+     * @param float $initialTtl The expiration delay of locks in seconds
```
Why `$initialTtl` is a constructor argument, whereas every other options are passed into `$options`?

kralos replied:
I'll move it to `$options`.

I've been thinking it would be simpler to remove the `drift` option in favour of asking the developer set a slightly higher lockTTL. This would both simplify the implementation and probably make more sense to the developer rather than trying to explain why drift is a separate option. I would have to document the danger of clock drift against this Store. What are your thoughts?

Commit:
#27345 removed drift option in favour or recommending setting TTL higher

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,251 @@
+     *
+     * A TTL index MUST BE used on MongoDB 2.2+ to automatically clean up expired locks.
+     *
+     *     db.lock.ensureIndex(
```
I suggest to provide a `public function createCollection` like in https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Traits/PdoTrait.php#L82 or https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php#L209 or https://github.com/symfony/symfony/pull/27456/files

kralos replied:
I've added a method to `createTTLIndex()`. There's no need in MongoDB to create the collection, they're created on the fly by the database.

Commit:
#27345 Added public createTTLIndex method

Commit:
#27345 Fixed coding standards

Commit:
#27345 Fixed coding standards

Commit:
#27345 Fixed coding standards

Commit:
#27345 Fixed coding standards

Commit:
#27345 Improved documentation

Commit:
#27345 Fixed spacing for fabbot

Commit:
#27345 reordered documentation in order of importance

Commit:
#27345 Fixed spacing for fabbot

Commit:
#27345 Fixed spacing for fabbot

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,276 @@
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+            'initialTtl' => 300.0,
```
I really wonder if options for lock should be merge with options for database connection?

See PdoTrait from cache components.

What is the opinion of other contributors?

kralos replied:
I originally had `initialTtl` as a separate constructor parameter, I brought it in to `$options` when you suggested `drift` should be.

I personally think it belongs outside of `$options`.

See also: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Adapter/PdoAdapter.php

jderusse commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,276 @@
+    private $collection;
+
+    /**
+     * @param \MongoDB\Client $mongo
```
Is there a way to pass a dns for lazy connection? See PdoTrait from cache components?

kralos replied:
The only way to avoid having to pass the database name as an options is to pass a `\MongoDB\Database` in `MongoDbStore::__construct` instead of a `\MongoDB\Client`.

A mongodb DSN does not contain a database to `use`. https://docs.mongodb.com/manual/reference/connection-string/

The `/database` component there is the database for authentication (unrelated to the database containing the data).

If I were to allow either a `\MongoDB\Database` or a `\MongoDB\Client`; the `database` option would then have to become mandatory only if a `\MongoDB\Client` is given.

kralos replied:
A `MongoDB\Client` instance is already lazy. I can turn off my database and still:
```php
$client = new \MongoDB\Client();
$store = new \Symfony\Component\Lock\Store\MongoDbStore($client, [
    'database' => 'test',
]);
```

only if I add something like:

```php
$store->createTTLIndex();
```

does it throw an exception.

Commit:
#27345 Updated return type and exception handling on createTTLIndex

Commit:
#27345 Updated exception handling on createTTLIndex

Commit:
#27345 Fixed spacing for fabbot

kralos commented on src/Symfony/Component/Lock/Store/MongoDbStore.php:
```diff
@@ -0,0 +1,276 @@
+     *
+     * @see http://docs.mongodb.org/manual/tutorial/expire-data/
+     */
+    public function createTTLIndex(): bool
```
My bad, I actually dug into this further and my return type `bool` was incorrect. I was reading the legacy driver documentation. The `createIndex` call actually does throw an exception if it fails. I have quickly tested this to verify and updated the phpdoc accordingly.

https://docs.mongodb.com/php-library/v1.3/reference/method/MongoDBCollection-createIndex/

Commit:
#27345 Removed initialTtl from  array and fixed unit test

Commit:
#27345 phpdoc adjusted for fabbot

Commit:
#27345 phpdoc adjusted for fabbot

Commit:
#27345 trigger ci retry

Commit:
#27345 trigger ci retry

nicolas-grekas commented:
@kralos would you mind squashing all commits in one please?

kralos commented:
Sorry, I broke this Pull request when squashing... has to open a new one.
